### ISSUE TITLE
Issues with Rust XGBoost Implementation - NaN F1 Score and Training Error with "objective" Hyper-parameter

### ISSUE THREAD

**Issue by cyccbxhl** [postgresml/postgresml#1622]
Reproduction Steps:
1. Start the official container using the following command:
`docker run \
       -it \
       -v postgresml_data:/var/lib/postgresql \
       -p 5433:5432 \
       -p 8000:8000 \
       ghcr.io/postgresml/postgresml:2.9.3 \
       sudo -u postgresml psql -d postgresml`

Issue 1: F1 Score is NaN
After executing the following SQL commands:
`SELECT pgml.load_dataset('digits');`
`SELECT * FROM pgml.train('Handwritten Digits', 'classification', 'pgml.digits', 'target', algorithm => 'xgboost', hyperparams => '{"n_estimators": 10}');`
The output:
<img width="1965" alt="image" src="https://github.com/user-attachments/assets/26427d15-6647-404c-9444-8463ccbf6fda">
The issue arises where the recall becomes NaN. This is due to the model trained by XGBoost classifying all test instances into the category "9," leading to TP + FN equals 0, hence causing division by zero in F1 score calculation. Notably, this issue does not surface when using runtime => python.

Issue 2: Training Error When Assigning "objective" Hyperparameter
Attempting to specify the "objective" hyperparam during training leads to an error. 
`SELECT * FROM pgml.train('Handwritten Digits', 'classification', 'pgml.digits', 'target', algorithm => 'xgboost', hyperparams => '{"n_estimators": 10, "objective": "multi:softmax"}');`
The error:
<img width="1355" alt="image" src="https://github.com/user-attachments/assets/c1e1aec4-fa4d-4e8b-a2ff-bf44cda57d17">
Similarly, there are no issues with the Python XGBoost.

**Comment by cyccbxhl** [comment]
The issue mentioned above, issue #2, has been identified as a problem where the output shapes differ depending on the chosen objective function, and the Rust XGBoost predict interface lacks adaptation for this variability.

Regarding issue #1, however, I am more concerned. I tested the rust-xgboost library your project depends on, merely running the example provided, and observed that regardless of how I adjusted hyperparameters such as eta, boost_rounds, etc., both the training and testing losses remained unchanged. This is certainly not in line with expectations, correct? I'm curious, during your initial integration of rust-xgboost, did you validate the precision and recall of the trained models? 
<img width="1255" alt="image" src="https://github.com/user-attachments/assets/615b22cd-f077-422d-a40a-72711f796e56">

**Comment by montanalow** [comment]
Thanks for the report, and digging into the underlying xgboost-rust implementation to hone in on the issue. I am able to reproduce your issue, and also confirm that adding the `runtime => 'python'` argument produces a better result.

```
INFO:  Metrics: {"f1": 0.93657863, "precision": 0.93709934, "recall": 0.9378532, "accuracy": 0.9376392, "mcc": 0.93084174, "fit_time": 1.7153733, "score_time": 0.006662062}
```

As you've pointed out, it's unlikely the bug is in pgml-extension, but I'll take a look and see if I can track down more info. Are you able to publish your test rust-xgboost code so that we can also report upstream?

**Comment by cyccbxhl** [comment]
> Thanks for the report, and digging into the underlying xgboost-rust implementation to hone in on the issue. I am able to reproduce your issue, and also confirm that adding the `runtime => 'python'` argument produces a better result.
> 
> ```
> INFO:  Metrics: {"f1": 0.93657863, "precision": 0.93709934, "recall": 0.9378532, "accuracy": 0.9376392, "mcc": 0.93084174, "fit_time": 1.7153733, "score_time": 0.006662062}
> ```
> 
> As you've pointed out, it's unlikely the bug is in pgml-extension, but I'll take a look and see if I can track down more info. Are you able to publish your test rust-xgboost code so that we can also report upstream?

Here 

`git clone --recursive https://github.com/postgresml/rust-xgboost.git`
`cd rust-xgboost/examples`

**Comment by cyccbxhl** [comment]
@montanalow I have resolved the NaN issue with the recall in the Rust version of XGBoost. I submitted a PR to the postgresml/rust-xgboost project: [Fix the NaN recall in xgboost training](https://github.com/postgresml/rust-xgboost/pull/12), Please review it.

**Comment by cyccbxhl** [comment]
@montanalow 
Regarding the two problems mentioned in this issue, I have submitted a PR： [1636](https://github.com/postgresml/postgresml/pull/1636), Please take a look and review it.
Thank you!
